src/libmscore/draw/bufferedpaintprovider.h

```diff
@@ -0,0 +1,154 @@
+        std::vector<QRectF> rects;
+        std::vector<QRectF> ellipses;
+        std::vector<FillPolygon> polygons;
+        std::vector<DrawText> texts;
+        std::vector<DrawRectText> rectTexts;
+        std::vector<DrawGlyphRun> glyphs;
+        std::vector<DrawPixmap> pixmaps;
+        std::vector<DrawTiledPixmap> tiledPixmap;
+    };
+
+    QRect window;
+    QRect viewport;
+    std::vector<Data> datas;
+};
+
+class BufferedPaintProvider : public IPaintProvider
+{
+public:
+    BufferedPaintProvider();
+
+    QPaintDevice* device() const override;
+    QPainter* qpainter() const override;
+
+    void begin(const std::string& name) override;
+    bool end(const std::string& name) override;
+    bool isActive() const override;
+
+    void setAntialiasing(bool arg) override;
+    void setCompositionMode(CompositionMode mode) override;
+
+    void setFont(const QFont& f) override;
+    const QFont& font() const override;
+
+    void setPen(const QPen& pen) override;
+    void setNoPen() override;
+    const QPen& pen() const override;
+
+    void setBrush(const QBrush& brush) override;
+    const QBrush& brush() const override;
+
+    void save() override;
+    void restore() override;
+
+    void setWorldTransform(const QTransform& matrix, bool combine = false) override;
+    const QTransform& worldTransform() const override;
+
+    void setTransform(const QTransform& transform, bool combine = false) override;
+    const QTransform& transform() const override;
+
+    void scale(qreal sx, qreal sy) override;
+    void rotate(qreal a) override;
+
+    void translate(const QPointF& offset) override;
+
+    QRect window() const override;
+    void setWindow(const QRect& window) override;
+    QRect viewport() const override;
+    void setViewport(const QRect& viewport) override;
+
+    // drawing functions
+    void fillPath(const QPainterPath& path, const QBrush& brush) override;
+    void drawPath(const QPainterPath& path) override;
+    void strokePath(const QPainterPath& path, const QPen& pen) override;
+
+    void drawLines(const QLineF* lines, int lineCount) override;
+
+    void drawRects(const QRectF* rects, int rectCount) override;
+
+    void drawEllipse(const QRectF& r) override;
+
+    void drawPolyline(const QPointF* points, int pointCount) override;
+
+    void drawPolygon(const QPointF* points, int pointCount, Qt::FillRule fillRule = Qt::OddEvenFill) override;
+    void drawConvexPolygon(const QPointF* points, int pointCount) override;
+
+    void drawText(const QPointF& p, const QString& s) override;
+    void drawText(const QRectF& r, int flags, const QString& text) override;
+
+    void drawGlyphRun(const QPointF& position, const QGlyphRun& glyphRun) override;
+
+    void fillRect(const QRectF& r, const QBrush& brush) override;
+
+    void drawPixmap(const QPointF& p, const QPixmap& pm) override;
+    void drawTiledPixmap(const QRectF& rect, const QPixmap& pm, const QPointF& offset = QPointF()) override;
```

Comment:
We use different namings in different methods for rects (r vs rect). I think it is better to use "rect" in all places

Reply:
fixed